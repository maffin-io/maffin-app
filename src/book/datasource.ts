import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';
import { useSWRConfig } from 'swr';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

// TODO: Need to mutate the accounts SWR key so Root is shown!
async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
