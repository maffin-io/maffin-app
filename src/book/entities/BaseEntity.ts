import {
  BaseEntity as BE,
  BeforeInsert,
  BeforeUpdate,
  PrimaryColumn,
  SaveOptions,
} from 'typeorm';
import * as v from 'class-validator';
import { QueryClient } from '@tanstack/react-query';

/**
 * Override BaseEntity so all entities have a generated guid in the format
 * we want and adds validation before saving or updating entities
 */
export default class BaseEntity extends BE {
  static CACHE_KEY = '';

  @BeforeInsert()
  @BeforeUpdate()
  async validate() {
    const errors = await v.validate(this, { stopAtFirstError: true });
    if (errors.length) {
      throw new Error(`validation failed: ${errors}`);
    }
  }

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.IsString()
    guid: string = crypto.randomUUID().substring(0, 31);

  async save(options?: SaveOptions): Promise<this> {
    const entity = await super.save(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity });
    }

    return entity;
  }

  async remove(options?: SaveOptions): Promise<this> {
    const entity = await super.remove(options);

    if (this.queryClient && ['Account', 'Commodity'].includes(this.constructor.name)) {
      updateCache({ queryClient: this.queryClient, entity, isDelete: true });
    }

    return entity;
  }

  get queryClient(): QueryClient | undefined {
    // @ts-ignore
    return this.constructor.dataSource.options.extra?.queryClient;
  }
}

/**
 * Updates `/api/<entity_name>` and
 * `/api/<entity_name>/<guid>` keys so changes are propagated
 * in the cache.
 *
 * - /api/<entity_name> is updated as a list of Entity[]
 * - /api/<entity_name>/<guid> is updated with the value of the Entity
 */
export async function updateCache(
  {
    queryClient,
    entity,
    isDelete = false,
  }: {
    queryClient: QueryClient,
    entity: BaseEntity,
    isDelete?: boolean,
  },
) {
  // @ts-ignore
  const key = entity.constructor.CACHE_KEY;

  queryClient.setQueryData(
    [key],
    (entities: BaseEntity[] | undefined) => {
      if (!entities) {
        return undefined;
      }

      const newEntities = [...entities];

      const index = entities.findIndex(e => e.guid === entity.guid);
      if (index === -1) { // New entity added
        newEntities.push(entity);
      } else if (!isDelete) { // Entity updated
        newEntities[index] = entity;
      } else if (isDelete) { // Entity deleted
        newEntities.splice(index, 1);
      }

      return newEntities;
    },
  );

  queryClient.setQueryData(
    [key, { guid: entity.guid }],
    !isDelete ? entity : null,
  );
}
