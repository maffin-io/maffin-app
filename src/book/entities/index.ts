export { default as Account } from './Account';
export { default as BankConfig } from './BankConfig';
export { default as Commodity } from './Commodity';
export { default as Price } from './Price';
export { default as Split } from './Split';
export { default as Transaction } from './Transaction';
export { default as Book } from './Book';
export { default as BaseEntity } from './BaseEntity';
