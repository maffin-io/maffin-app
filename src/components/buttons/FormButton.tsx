import React from 'react';

import { DataSourceContext } from '@/hooks';
import Modal, { ModalRef } from '@/components/ui/Modal';

export interface FormButtonProps extends Omit<
React.ButtonHTMLAttributes<HTMLButtonElement>,
'children'
> {
  buttonContent: string | React.ReactElement,
  modalTitle: string,
  children: React.ReactElement,
}

export default function FormButton({
  buttonContent,
  modalTitle,
  children,
  className = 'btn btn-primary',
  ...props
}: FormButtonProps): React.JSX.Element {
  const { save } = React.useContext(DataSourceContext);
  const modalRef = React.useRef<ModalRef>(null);

  return (
    <Modal
      ref={modalRef}
      className="modal card"
      triggerContent={buttonContent}
      triggerProps={{ ...props, className }}
      showClose
    >
      <span>{modalTitle}</span>
      <div>
        {
          React.cloneElement(
            children,
            {
              // @ts-ignore
              onSave: (e: unknown) => {
                // @ts-ignore
                children.props.onSave?.(e);
                save();
                modalRef.current?.closeModal();
              },
            },
          )
        }
      </div>
    </Modal>
  );
}
