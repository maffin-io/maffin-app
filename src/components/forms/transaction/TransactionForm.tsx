import React from 'react';
import { DateTime } from 'luxon';
import { useForm } from 'react-hook-form';
import { classValidatorResolver } from '@hookform/resolvers/class-validator';
import { mutate } from 'swr';
import { IsNull } from 'typeorm';
import { Tooltip } from 'react-tooltip';

import {
  Split,
  Transaction,
} from '@/book/entities';
import { isInvestment } from '@/book/helpers/accountType';
import SplitsField from './SplitsField';
import type { FormValues } from './types';

const resolver = classValidatorResolver(Transaction, { validator: { stopAtFirstError: true } });

export type TransactionFormProps = {
  action?: 'add' | 'update' | 'delete',
  onSave: Function,
  defaultValues?: FormValues,
};

export default function TransactionForm({
  action = 'add',
  onSave,
  defaultValues,
}: TransactionFormProps): JSX.Element {
  const form = useForm<FormValues>({
    defaultValues,
    mode: 'onChange',
    resolver,
  });

  const { errors } = form.formState;

  let disabled = false;
  let submitButtonText = 'Save';
  let submitButtonClass = 'btn-primary';
  if (action === 'update') {
    submitButtonText = 'Update';
    submitButtonClass = 'btn-warn';
  }
  if (action === 'delete') {
    disabled = true;
    submitButtonText = 'Delete';
    submitButtonClass = 'btn-danger';
  }

  return (
    <form onSubmit={form.handleSubmit((data) => onSubmit(data, action, onSave))}>
      <fieldset className="text-sm my-5">
        <label htmlFor="dateInput" className="inline-block mb-2">Date</label>
        <input
          id="dateInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('date')}
          type="date"
        />
        <p className="invalid-feedback">{errors.date?.message}</p>
      </fieldset>

      <fieldset className="text-sm my-5">
        <label htmlFor="descriptionInput" className="inline-block mb-2">Description</label>
        <span
          className="badge"
          data-tooltip-id="description-help"
        >
          ?
        </span>
        <Tooltip
          id="description-help"
          className="bg-cyan-600 w-1/3 text-white rounded-lg p-2"
          disableStyleInjection
        >
          <p className="mb-2">
            Add a meaningful description to your transaction. The description is shown in the
            list of transactions in the account detail page.
          </p>
          <p>
            It helps to set the same description
            for similar transactions. For example, for a Groceries transaction, you can set
            &quot;Supermarket&quot; to all of them.
          </p>
        </Tooltip>
        <input
          id="descriptionInput"
          disabled={disabled}
          className="block w-full m-0"
          {...form.register('description')}
          type="text"
          placeholder="Enter description"
        />
        <p className="invalid-feedback">{errors.description?.message}</p>
      </fieldset>

      <fieldset className="text-sm my-5">
        <label className="inline-block mb-2">Splits</label>
        <SplitsField
          disabled={disabled}
          form={form}
        />
      </fieldset>

      <input
        {...form.register(
          'guid',
        )}
        hidden
        type="text"
      />

      <input
        {...form.register(
          'fk_currency',
        )}
        hidden
        type="number"
      />

      <div className="flex w-full gap-2 items-center justify-center">
        <button className={submitButtonClass} type="submit" disabled={Object.keys(errors).length > 0}>
          {submitButtonText}
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, action: 'add' | 'update' | 'delete', onSave: Function) {
  const transaction = Transaction.create({
    ...data,
    guid: data.guid || undefined,
    date: DateTime.fromISO(data.date),
  });

  if (action === 'add') {
    await transaction.save();
  } else if (action === 'update') {
    await transaction.save();
    await Split.delete({
      fk_transaction: IsNull(),
    });
  } else if (action === 'delete') {
    // Not using cascade here because seems it has problems with
    // old data. Deleting splits manually for now
    await Split.remove(transaction.splits);
    await transaction.remove();
  }

  transaction.splits.forEach(split => {
    if (isInvestment(split.account)) {
      mutate('/api/investments');
    }
    mutate(`/api/splits/${split.account.guid}`);
  });

  mutate('/api/txs/latest');
  onSave();
}
