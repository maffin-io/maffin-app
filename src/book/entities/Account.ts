import * as v from 'class-validator';
import {
  Column,
  Entity, JoinColumn,
  ManyToOne,
  OneToMany,
  RelationId,
  Tree,
  TreeParent,
  TreeChildren,
  SaveOptions,
} from 'typeorm';
import type { QueryClient } from '@tanstack/react-query';

import {
  getAllowedSubAccounts,
  ASSET_ACCOUNTS,
  LIABILITY_ACCOUNTS,
} from '@/book/helpers/accountType';
import type Commodity from './Commodity';
import Split from './Split';
import BaseEntity from './BaseEntity';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static CACHE_KEY = ['api', 'accounts'];
  static TYPES = [
    'ROOT',
    'EQUITY',
    'INCOME',
    'EXPENSE',
    ...ASSET_ACCOUNTS,
    ...LIABILITY_ACCOUNTS,
  ];

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.Matches(/[a-zA-Z.]+/)
  @v.Length(1, 2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @CheckAccountType()
  @v.IsIn(Account.TYPES)
    type!: string;

  path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  @v.IsNotEmpty({ message: 'parent is required' })
  @v.ValidateIf(o => o.type !== 'ROOT')
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @RelationId((account: Account) => account.parent)
    parentId: string;

  @TreeChildren()
    children!: Account[];

  @RelationId((account: Account) => account.children)
    childrenIds: string[];

  @ManyToOne('Commodity', { eager: true, cascade: true })
  @JoinColumn({ name: 'commodity_guid' })
  @CheckIECommodity()
  @v.ValidateIf(o => o.type !== 'ROOT')
  @v.IsNotEmpty({ message: 'commodity is required' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    nullable: true,
  })
  @v.IsOptional()
  @v.Length(4, 2048)
    description?: string;

  @Column({
    default: false,
  })
    hidden!: boolean;

  @CheckPlaceholder()
  @Column({
    default: false,
  })
    // Placeholders are hierarchical accounts that don't containg
    // transactions, they are just parents of other accounts
    placeholder!: boolean;

  async save(options?: SaveOptions): Promise<this> {
    const account = await super.save(options);

    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    return account;
  }

  async remove(options?: SaveOptions): Promise<this> {
    if (this.queryClient) {
      updateCache({ queryClient: this.queryClient });
    }

    const account = await super.remove(options);
    return account;
  }
}

export async function updateCache({ queryClient }: { queryClient: QueryClient }) {
  queryClient.invalidateQueries({
    queryKey: [...Account.CACHE_KEY],
  });
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });

/**
 * Checks allowed types for parent account given the current account's type
 */
function CheckAccountType(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkAccountType',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(type: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.parent) {
            return getAllowedSubAccounts(account.parent.type).includes(type);
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          const allowedTypes = getAllowedSubAccounts(account.parent.type);

          return `only ${allowedTypes} types can be selected with ${account.parent.type} account as parent`;
        },
      },
    });
  };
}

/**
 * Checks that income/expense commodity is the same as the parent.
 *
 * This is a way to check that we always keep the same currency for income/expense
 * account as we only support them in the main currency.
 */
function CheckIECommodity(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkIECommodity',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        validate(fk_commodity: Commodity, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (
            account.parent
            && account.parent.type !== 'ROOT'
            && ['INCOME', 'EXPENSE'].includes(account.type)
          ) {
            return fk_commodity.guid === (account.parent.fk_commodity as Commodity).guid;
          }

          return true;
        },

        defaultMessage(args: v.ValidationArguments) {
          const account = args.object as Account;
          return `Income and Expense accounts must have ${(account.parent.fk_commodity as Commodity).mnemonic} as their commodity`;
        },
      },
    });
  };
}

/**
 * Checks if an account can be a placeholder. Accounts can be Placeholders
 * if they don't have any transactions associated.
 */
function CheckPlaceholder(validationOptions?: v.ValidationOptions) {
  return function f(object: Account, propertyName: string) {
    v.registerDecorator({
      name: 'checkPlaceholder',
      target: object.constructor,
      propertyName,
      options: validationOptions,
      validator: {
        async validate(_: string, args: v.ValidationArguments) {
          const account = args.object as Account;
          if (account.guid) {
            const [, numSplits] = await Split.findAndCount({
              where: { fk_account: { guid: account.guid } },
            });

            if (numSplits > 0 && account.placeholder) {
              return false;
            }
          }

          return true;
        },

        defaultMessage() {
          return 'Placeholder accounts cannot have transactions';
        },
      },
    });
  };
}
