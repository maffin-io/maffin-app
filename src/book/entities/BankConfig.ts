import {
  Column,
  Entity,
  OneToMany,
} from 'typeorm';
import {
  IsString,
} from 'class-validator';

import BaseEntity from './BaseEntity';
import type Account from './Account';

@Entity('bank_config')
export default class BankConfig extends BaseEntity {
  @Column({
    type: 'text',
    length: 2048,
  })
  @IsString()
    token!: string;

  @OneToMany(
    'Account',
    (account: Account) => account.config,
  )
    accounts!: Account[];
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(BankConfig, 'name', { value: 'BankConfig' });
