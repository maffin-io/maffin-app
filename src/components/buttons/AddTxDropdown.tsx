import React from 'react';
import { DateTime } from 'luxon';
import { BiPlusCircle, BiShareAlt } from 'react-icons/bi';

import FormButton from '@/components/buttons/FormButton';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Split } from '@/book/entities';
import type { Account } from '@/book/entities';

export type AddTxDropdownProps = {
  account: Account,
  latestDate: DateTime,
};

export default function AddTxDropdown({
  account,
  latestDate,
}: AddTxDropdownProps): JSX.Element {
  return (
    <div className="group relative h-full">
      <button
        type="button"
        className="group-hover:bg-cyan-700/80 dark:group-hover:bg-cyan-600 flex h-full w-full items-center btn btn-primary"
      >
        <BiPlusCircle className="mr-1" />
        Add
      </button>
      <ul className="absolute rounded-md w-40 hidden py-2 group-hover:block bg-white dark:bg-dark-800 ">
        <li className="text-sm hover:bg-light-100 dark:hover:bg-dark-700">
          <FormButton
            id="add-tx"
            className="flex items-center text-left px-3 py-2 w-full text-cyan-700 hover:text-cyan-600 whitespace-nowrap"
            modalTitle={`Add transaction to ${account.name}`}
            buttonContent={(
              <>
                <BiShareAlt className="mr-1" />
                Transaction
              </>
            )}
          >
            <TransactionForm
              defaultValues={
                {
                  date: (latestDate || DateTime.now()).toISODate() as string,
                  description: '',
                  splits: [Split.create({ fk_account: account }), new Split()],
                  fk_currency: account.commodity,
                }
              }
            />
          </FormButton>
        </li>
      </ul>
    </div>
  );
}
