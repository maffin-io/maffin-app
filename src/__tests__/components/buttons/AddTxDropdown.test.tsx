import React from 'react';
import { DateTime } from 'luxon';
import { render, screen } from '@testing-library/react';

import AddTxDropdown from '@/components/buttons/AddTxDropdown';
import FormButton from '@/components/buttons/FormButton';
import TransactionForm from '@/components/forms/transaction/TransactionForm';
import { Account, Split } from '@/book/entities';

jest.mock('@/components/buttons/FormButton', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="FormButton">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/transaction/TransactionForm', () => jest.fn(
  () => <div data-testid="TransactionForm" />,
));

describe('AddTxDropdown', () => {
  beforeEach(() => {
    jest.spyOn(Split, 'create').mockReturnValue({ guid: 'createdSplit' } as Split);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', async () => {
    const account = {
      guid: 'guid',
      path: 'path',
      name: 'account',
      type: 'TYPE',
      parentId: 'parent',
      commodity: {
        mnemonic: 'EUR',
      },
    } as Account;

    const { container } = render(<AddTxDropdown account={account} latestDate={DateTime.now()} />);

    await screen.findAllByTestId('FormButton');
    expect(FormButton).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        id: 'add-tx',
        modalTitle: 'Add transaction to account',
      }),
      {},
    );
    expect(TransactionForm).toBeCalledWith(
      {
        defaultValues: {
          date: DateTime.now().toISODate(),
          description: '',
          fk_currency: {
            mnemonic: 'EUR',
          },
          splits: [
            { guid: 'createdSplit' },
            {
              action: '',
              guid: expect.any(String),
            },
          ],
        },
      },
      {},
    );

    expect(container).toMatchSnapshot();
  });
});
