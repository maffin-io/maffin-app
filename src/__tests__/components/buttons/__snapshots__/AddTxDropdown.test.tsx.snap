// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTxDropdown renders as expected 1`] = `
<div>
  <div
    class="group relative h-full"
  >
    <button
      class="group-hover:bg-cyan-700/80 dark:group-hover:bg-cyan-600 flex h-full w-full items-center btn btn-primary"
      type="button"
    >
      <svg
        class="mr-1"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
        />
        <path
          d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
        />
      </svg>
      Add
    </button>
    <ul
      class="absolute rounded-md w-40 hidden py-2 group-hover:block bg-white dark:bg-dark-800 "
    >
      <li
        class="text-sm hover:bg-light-100 dark:hover:bg-dark-700"
      >
        <div
          data-testid="FormButton"
        >
          <div
            data-testid="TransactionForm"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
